kernel/src/log_segment.rs

```diff
@@ -4,20 +4,42 @@
 use crate::actions::{get_log_schema, Metadata, Protocol, METADATA_NAME, PROTOCOL_NAME};
 use crate::path::ParsedLogPath;
 use crate::schema::SchemaRef;
-use crate::{DeltaResult, Engine, EngineData, Error, Expression, ExpressionRef, FileMeta};
-
+use crate::snapshot::CheckpointMetadata;
+use crate::utils::require;
+use crate::{
+    DeltaResult, Engine, EngineData, Error, Expression, ExpressionRef, FileSystemClient, Version,
+};
 use itertools::Itertools;
+use std::cmp::Ordering;
 use std::sync::{Arc, LazyLock};
+use tracing::warn;
 use url::Url;
 
+/// A [`LogSegment`] represents a contiguous section of the log and is made up of checkpoint files
+/// and commit files. It is built with [`LogSegmentBuilder`], and guarantees the following:
+///     1. Commit/checkpoint file versions will be less than or equal to `end_version` if specified.
+///     2. Commit file versions will be greater than or equal to `start_version` if specified.
+///     3. If checkpoint(s) is/are present in the range, only commits with versions greater than the most
+///        recent checkpoint version are retained. Checkpoints can be omitted (and this rule skipped)
+///        whenever the `LogSegment` is created. See [`LogSegmentBuilder::with_omit_checkpoint_files`].
+///
+/// [`LogSegment`] is used in both  [`Snapshot`] and in `TableChanges` to hold commit files and
+/// checkpoint files.
+/// - For a Snapshot at version `n`: Its LogSegment is made up of zero or one checkpoint, and all
+///   commits between the checkpoint and the end version `n`.
+/// - For a TableChanges between versions `a` and `b`: Its LogSegment is made up of zero
+/// checkpoints and all commits between versions `a` and `b`
+///
+/// [`Snapshot`]: crate::snapshot::Snapshot
 #[derive(Debug)]
 #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
 pub(crate) struct LogSegment {
+    pub end_version: Version,
     pub log_root: Url,
-    /// Reverse order sorted commit files in the log segment
+    /// Commit files in the log segment
     pub commit_files: Vec<ParsedLogPath>,
```

Comment:
nit:
```suggestion
    /// Sorted commit files in the log segment (see builder for sorting)
    pub commit_files: Vec<ParsedLogPath>,
```

and probably in the future we want to propagate sort information down to there but let's take that as follow-up